dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ContainerClassLoader.java

```diff
@@ -461,6 +426,212 @@ private static abstract class AbstractUniversalContainer<E> implements Universal
+                try {
+                    containerFile = resourceContainerURL != null ? new File(resourceContainerURL.toURI()) : null;
+                } catch (URISyntaxException e) {
+                    // Auto-FFDC
+                }
+                resourceContainerDir = containerFile != null && isDirectory(containerFile) ? containerFile : null;
+                resourceSharedClassCacheURL = hook != null ? createSharedClassCacheURL(resourceContainerURL, originalRoot, resourceContainerDir) : null;
+            }
+        }
+
+        URL createContainerURL(URL base) {
+            try {
+                if ("file".equals(base.getProtocol())) {
+                    // use file URLs as-is
+                    return base;
+                }
+
+                URLConnection conn = base.openConnection();
+                if (conn instanceof JarURLConnection) {
+                    return ((JarURLConnection) conn).getJarFileURL();
+                } else if (conn instanceof WSJarURLConnection) {
+                    return ((WSJarURLConnection) conn).getFile().toURI().toURL();
+                }
+                throw new UnsupportedOperationException(base.getProtocol());
+            } catch (IOException err) {
+                throw new RuntimeException(err);
+            }
+        }
+
+        URL createSharedClassCacheURL(URL containerURL, URL originalRoot, File containerDir) {
+            if (containerURL == null) {
+                return null;
+            }
+            if (containerDir != null) {
+                return containerURL;
+            }
+            try {
+                String containerPath = containerURL.getPath();
+                if (containerPath.endsWith(".jar") || containerPath.endsWith(".zip")) {
+                    // use containerURL as-is if it is a jar or zip extension
+                    return containerURL;
+                }
+
+                String basePath = originalRoot.getPath();
+                int bangSlash = basePath.lastIndexOf("!/");
+                if (bangSlash >= 0) {
+                    // append the original !/ path (likely !/WEB-INF/classes)
+                    return new URL(containerURL.toExternalForm() + basePath.substring(bangSlash));
+                }
+                // If the URL is not to a directory and does not end ith jar or zip file extension then
+                // we assume it is still some type of archive (e.g. rar).
+                // The Semeru shared classes cache logic will not recognize the URL as a valid archive
+                // if it does not end with jar or zip file extension.
+                // The URL will get recognized as a valid archive if it does contain '!/' with any path after.
+                // Here we append '!/l' to the URL so that the Semeru shared classes cache logic will treat it
+                // as a valid archive.  For example: file://path/to/myResourceAdaptor.rar!/l
+                return new URL(containerURL + "!/l");
```

Comment:
```suggestion
                return new URL(containerURL.toExternalForm() + "!/l");
```
To be consistent with the code above this section.